### PR TITLE
MP/4.Add username information to SignUp and Login

### PR DESCRIPTION
**Pull request by MPucheta**
- Added a bottom tab navigator for future screens, with this change I may be able to do the screens independently (To continue working with them as separate features instead of coding on top of the previous branch, and also reduce the number of different files I change per pull). **As for now, this pull must be merged after MP/3**

- Added Username to SignUp and Login. Now we can login with the username too (and as such, at the moment, the usernames must also be unique).

- I did quite a few tests with the database as I will probably use it later and I wanted to do an abstraction, regardless of this, the code retains some verbosity. A refactoring may be useful at the end of the iteration.

**Changes after refactoring**

- **Screens are now separated into more granular components, each with their own styles. Some things could be shared, such as Background and AuthenticatingIndicator, others such as Headers and AccountLink/NoAccountLink can be abstracted but in doing so they result confusing to read.**

- **Changed database to cloud firestore (noSQL based on documents)**

- **I had an urge of abstracting firebase.auth(), firebase.firestore() and initialization into a single file for modularity. It seemed a little overkill as the firebase API is very clear to use and read and another layer could have the adversary effect, what do you think?**

- I attach some pics

image

image

### CODE REVIEW

File: src/databaseService/index.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+const set = async (refPath, data) => {
+  try {
+    await firebase.database().ref(refPath).set(data)
+  } catch (error) {
+    console.log(error.message)
+  }
+}
+
+const get = async (refPath, mode = 'value') => {
+  let values = null
```
Comment (abelosorio) [inline src/databaseService/index.js]: I think you don't need this variable. You can just return `databaseSnapshot.val()`.

File: src/databaseService/index.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  return json[Object.keys(json)[0]].email
+}
+
+const queryEmailFromUsername = async (username, refPath = 'users/') => {
+  const ref = firebase.database().ref(refPath)
+  let email = null
+  try {
+    const databaseSnapshot = await ref.orderByChild('username').equalTo(username).once('value')
+    const json = databaseSnapshot.val()
+    email = json && getEmailFromJson(json)
```
Comment (abelosorio) [inline src/databaseService/index.js]: Same here, you can do instead: `return json && getEmailFromJson(json)`

File: src/databaseService/index.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  } catch (error) {
+    console.log(error.message)
+  }
+  return email
+}
+
+const usernameAlreadyInUse = async (username, refPath = 'users/') => {
+  const email = await queryEmailFromUsername(username, refPath)
+  return email !== null
```
Comment (abelosorio) [inline src/databaseService/index.js]: Let's leave a blank line before return statements (when they are among other statements within the same block).

File: src/screens/authentication/Login/index.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
+import React, { useState } from 'react'
```
Comment (abelosorio) [inline src/screens/authentication/Login/index.js:1]: The Login and Signup components are quite long. I'd suggest to split them out into smaller and reusable files. For example:
```javascript
    <LoginBackground>
      <LoginHeader />
      <LoginForm
        initialValues={initialValues}
        errors={errors}
        authenticating={authenticating}
        onSubmit={handleSubmit}
      />
      <NoAccountLink />
    </LoginBackground>
```

File: src/screens/authentication/utils/index.js [hunk 5]
```diff
@@ -0,0 +1 @@
+export const isNotEmail = (text) => !text.includes('@')
```
Comment (abelosorio) [inline src/screens/authentication/utils/index.js]: A very small detail, but I had to come here to see what `isNotEmail` was. I'd suggest renaming to `isValidEmail`. @nahudalla what do you say?

**Review by abelosorio — COMMENTED** [review]
Awesome!! It's coming along! A few small comments, tho.

File: src/screens/authentication/utils/index.js (hunk 5 above)
Comment (nahudalla) [reply src/screens/authentication/utils/index.js]: Yes, I agree. Also, this is not a correct way of checking if it's not a valid email address... but I think it's OK for now

**Review by nahudalla — APPROVED** [review]

File: src/screens/authentication/utils/index.js (hunk 5 above)
Comment (MPucheta) [reply src/screens/authentication/utils/index.js]: You are correct, I considered if putting a RegExp was necessary as a badly formatted email would not be logged by firebase and only needed the distinction between username and email. 

I will rethink if what I intended is a "**isUsername**" functionality instead.

**Review by nahudalla — APPROVED** [review]

**Review by abelosorio — APPROVED** [review]

**Pull request merged**